### PR TITLE
Chooser -> subsystems-dev

### PR DESCRIPTION
**Pull request by drorta**

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/chooser/Chooser.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+		this.commandsBuilder = new ChooserCommandsBuilder(this);
+
+		updateInputs();
+	}
+
+	public ChooserCommandsBuilder getCommandsBuilder() {
+		return commandsBuilder;
+	}
+
+	public boolean isObjectIn() {
```
Comment (YHerm) [inline src/main/java/frc/robot/subsystems/chooser/Chooser.java]:
```suggestion
	public boolean isNoteIn() {
```

File: src/main/java/frc/robot/subsystems/chooser/ChooserCommandsBuilder.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+	//@formatter:off
+	public Command setPower(double power) {
+        return new FunctionalCommand(
+				() -> {},
+				() -> chooser.setPower(power),
+				interrupted -> chooser.stop(),
+				() -> false,
+				chooser
+		).withName("Set chooser power: " + power);
```
Comment (YHerm) [inline src/main/java/frc/robot/subsystems/chooser/ChooserCommandsBuilder.java]:
```suggestion
		).withName("Set power: " + power);
```

File: src/main/java/frc/robot/subsystems/chooser/ChooserCommandsBuilder.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+				() -> chooser.setPower(power.getAsDouble()),
+				interrupted -> chooser.stop(),
+				() -> false,
+				chooser
+		).withName("Set chooser power by supplier");
+    }
+	//@formatter:on
+
+	public Command stop() {
+		return new RunCommand(chooser::stop, chooser).withName("Stop chooser");
```
Comment (YHerm) [inline src/main/java/frc/robot/subsystems/chooser/ChooserCommandsBuilder.java]:
```suggestion
		return new RunCommand(chooser::stop, chooser).withName("Stop");
```

File: src/main/java/frc/robot/subsystems/chooser/ChooserCommandsBuilder.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    public Command setPower(DoubleSupplier power) {
+        return new FunctionalCommand(
+				() -> {},
+				() -> chooser.setPower(power.getAsDouble()),
+				interrupted -> chooser.stop(),
+				() -> false,
+				chooser
+		).withName("Set chooser power by supplier");
```
Comment (YHerm) [inline src/main/java/frc/robot/subsystems/chooser/ChooserCommandsBuilder.java]:
```suggestion
		).withName("Set power by supplier");
```

File: src/main/java/frc/robot/subsystems/chooser/factory/RealChooserConstants.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+	private final static double DEBOUNCE_TIME_SECONDS = 0.05;
+	private final static Debouncer.DebounceType DEBOUNCE_TYPE = Debouncer.DebounceType.kBoth;
+	private final static SparkLimitSwitch.Type REVERSE_LIMIT_SWITCH_TYPE = SparkLimitSwitch.Type.kNormallyOpen;
+
+	public static ChooserStuff generateChooserStuff(String logPath) {
+		SparkMaxWrapper sparkMAXWrapper = new SparkMaxWrapper(IDs.CANSparkMAXIDs.CHOOSER);
+		SysIdRoutine.Config config = new SysIdRoutine.Config();
+		BrushlessSparkMAXMotor motor = new BrushlessSparkMAXMotor(logPath, sparkMAXWrapper, config);
+
+		Supplier<Double> voltage = () -> sparkMAXWrapper.getBusVoltage() * sparkMAXWrapper.getAppliedOutput();
```
Comment (YHerm) [inline src/main/java/frc/robot/subsystems/chooser/factory/RealChooserConstants.java]: use get voltage

File: src/main/java/frc/robot/subsystems/chooser/factory/RealChooserConstants.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+		SysIdRoutine.Config config = new SysIdRoutine.Config();
+		BrushlessSparkMAXMotor motor = new BrushlessSparkMAXMotor(logPath, sparkMAXWrapper, config);
+
+		SuppliedDoubleSignal voltageSignal = new SuppliedDoubleSignal("voltage", sparkMAXWrapper::getVoltage);
+
+		Supplier<Double> position = () -> sparkMAXWrapper.getEncoder().getPosition();
+		SuppliedAngleSignal positionSignal = new SuppliedAngleSignal("position", position, AngleUnit.ROTATIONS);
+
+		BooleanSupplier isBeamBroken = () -> sparkMAXWrapper.getReverseLimitSwitch(REVERSE_LIMIT_SWITCH_TYPE).isPressed();
+		sparkMAXWrapper.getReverseLimitSwitch(REVERSE_LIMIT_SWITCH_TYPE).enableLimitSwitch(false);
```
Comment (YHerm) [inline src/main/java/frc/robot/subsystems/chooser/factory/RealChooserConstants.java]: and current and conversion factor

File: src/main/java/frc/robot/subsystems/chooser/Chooser.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+public class Chooser extends GBSubsystem {
+
+	private final IMotor motor;
+	private final ChooserStuff chooserStuff;
+	private final ChooserCommandsBuilder commandsBuilder;
+	private final IDigitalInput digitalInput;
+	private final DigitalInputInputsAutoLogged digitalInputInputs;
+
+	public Chooser(ChooserStuff chooserStuff) {
```
Comment (YHerm) [inline src/main/java/frc/robot/subsystems/chooser/Chooser.java:17]: organize like in funnel

**Review by YHerm — APPROVED** [review]

File: src/main/java/frc/robot/subsystems/chooser/factory/RealChooserConstants.java [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+import java.util.function.Supplier;
+
+public class RealChooserConstants {
+
+	private final static double DEBOUNCE_TIME_SECONDS = 0.05;
+	private final static Debouncer.DebounceType DEBOUNCE_TYPE = Debouncer.DebounceType.kBoth;
+	private final static SparkLimitSwitch.Type REVERSE_LIMIT_SWITCH_TYPE = SparkLimitSwitch.Type.kNormallyOpen;
+
+	private final static int CURRENT_LIMIT = 30;
+	private final static int GEAR_RATIO = 1;
```
Comment (YHerm) [inline src/main/java/frc/robot/subsystems/chooser/factory/RealChooserConstants.java]: no need just put it

File: src/main/java/frc/robot/subsystems/chooser/Chooser.java [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+	public void updateInputs() {
+		digitalInput.updateInputs(digitalInputInputs);
+		motor.updateSignals(chooserStuff.voltageSignal(), chooserStuff.positionSignal());
+	}
+
+	@Override
+	protected void subsystemPeriodic() {
+		updateInputs();
+		Logger.processInputs(chooserStuff.digitalInputLogPath(), digitalInputInputs);
```
Comment (YHerm) [inline src/main/java/frc/robot/subsystems/chooser/Chooser.java]: add record output to isNoteIn

File: src/main/java/frc/robot/subsystems/chooser/factory/RealChooserConstants.java [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+	private final static double DEBOUNCE_TIME_SECONDS = 0.05;
+	private final static Debouncer.DebounceType DEBOUNCE_TYPE = Debouncer.DebounceType.kBoth;
+	private final static SparkLimitSwitch.Type REVERSE_LIMIT_SWITCH_TYPE = SparkLimitSwitch.Type.kNormallyOpen;
+
+	public static ChooserStuff generateChooserStuff(String logPath) {
+		SparkMaxWrapper sparkMAXWrapper = new SparkMaxWrapper(IDs.CANSparkMAXIDs.CHOOSER);
+		SysIdRoutine.Config config = new SysIdRoutine.Config();
+
+		sparkMAXWrapper.setSmartCurrentLimit(30);
```
Comment (YHerm) [inline src/main/java/frc/robot/subsystems/chooser/factory/RealChooserConstants.java:26]: add setIdleMode coast

File: src/main/java/frc/robot/subsystems/chooser/ChooserCommandsBuilder.java [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+    public Command setPower(DoubleSupplier power) {
+        return new FunctionalCommand(
+				() -> {},
+				() -> chooser.setPower(power.getAsDouble()),
+				interrupted -> chooser.stop(),
+				() -> false,
+				chooser
+		).withName("Set power by supplier");
+    }
+	//@formatter:on
```
Comment (YHerm) [inline src/main/java/frc/robot/subsystems/chooser/ChooserCommandsBuilder.java]: move below stop

File: src/main/java/frc/robot/constants/IDs.java [hunk 12]
```diff
@@ -11,4 +13,10 @@ public class IDs {
 		PowerDistribution.ModuleType.kRev
 	);
 
+	public static class CANSparkMAXIDs {
+
+		public static final SparkMaxDeviceID CHOOSER = new SparkMaxDeviceID(1, CANSparkLowLevel.MotorType.kBrushless);
```
Comment (YHerm) [inline src/main/java/frc/robot/constants/IDs.java]:
```suggestion
		public static final SparkMaxDeviceID CHOOSER = new SparkMaxDeviceID(2, CANSparkLowLevel.MotorType.kBrushless);
```

**Review by YHerm — DISMISSED** [review]

**Review by hillelv — DISMISSED** [review]
just remove the setBrake

File: src/main/java/frc/robot/Robot.java [hunk 13]
```diff
@@ -16,7 +19,11 @@ public class Robot {
 
 	public static final RobotType ROBOT_TYPE = RobotType.determineRobotType();
 
+	private final Chooser chooser;
+
 	public Robot() {
+		this.chooser = new Chooser(ChooserFactory.create(ChooserConstants.LOG_PATH));
```
Comment (YHerm) [inline src/main/java/frc/robot/Robot.java:25]: Add get function

File: src/main/java/frc/robot/subsystems/chooser/factory/ChooserFactory.java [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+
+import frc.robot.Robot;
+import frc.robot.subsystems.chooser.ChooserStuff;
+
+public class ChooserFactory {
+
+	public static ChooserStuff create(String logPath) {
+		return switch (Robot.ROBOT_TYPE) {
+			case REAL -> RealChooserConstants.generateChooserStuff(logPath);
+			case SIMULATION -> RealChooserConstants.generateChooserStuff(logPath);
```
Comment (YHerm) [inline src/main/java/frc/robot/subsystems/chooser/factory/ChooserFactory.java]:
```suggestion
			case SIMULATION -> null;
```

File: src/main/java/frc/robot/subsystems/chooser/Chooser.java [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+	public void updateInputs() {
+		digitalInput.updateInputs(digitalInputInputs);
+		motor.updateSignals(chooserStuff.voltageSignal());
+	}
+
+	@Override
+	protected void subsystemPeriodic() {
+		updateInputs();
+		Logger.processInputs(chooserStuff.digitalInputLogPath(), digitalInputInputs);
+		Logger.recordOutput(chooserStuff.digitalInputLogPath() + "isNoteIn", digitalInputInputs.debouncedValue);
```
Comment (YHerm) [inline src/main/java/frc/robot/subsystems/chooser/Chooser.java]: use the function

File: src/main/java/frc/robot/subsystems/chooser/Chooser.java [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+
+	public void updateInputs() {
+		digitalInput.updateInputs(digitalInputInputs);
+		motor.updateSignals(chooserStuff.voltageSignal());
+	}
+
+	@Override
+	protected void subsystemPeriodic() {
+		updateInputs();
+		Logger.processInputs(chooserStuff.digitalInputLogPath(), digitalInputInputs);
```
Comment (YHerm) [inline src/main/java/frc/robot/subsystems/chooser/Chooser.java]: move to update

**Review by YHerm — APPROVED** [review]

**Review by hillelv — APPROVED** [review]

**Review by YoniKiriaty — APPROVED** [review]

**Pull request closed without merging**